exec/testing/fake_exec.go

```diff
@@ -45,6 +71,82 @@ func (fake *FakeExec) LookPath(file string) (string, error) {
 	return fake.LookPathFunc(file)
 }
 
+func getCaller(n int) string {
+	_, file, line, ok := runtime.Caller(n + 1)
+	if !ok {
+		return "???"
+	}
+	if i := strings.LastIndexAny(file, "/\\"); i >= 0 {
+		file = file[i+1:]
+	}
+	return fmt.Sprintf("%s:%d", file, line)
+}
+
+func (fake *FakeExec) fail(format string, args ...interface{}) {
+	msg := fmt.Sprintf(format, args...)
+	if fake.T == nil {
+		panic(msg)
+	}
+
+	// Don't re-fail from a "defer fexec.AssertExpectedCommands()" if we already failed once
+	if fake.T.Failed() {
+		return
+	}
+	fake.T.Fatal(msg)
+}
+
+// AddCommand adds a Cmd to be returned by a future fake.Command(...) call. Call
+// SetCombinedOutput or SetRunOutput on the result to specify its result.
+func (fake *FakeExec) AddCommand(cmd string, args ...string) *FakeCmd {
```

Comment:
I like that this makes things easier, but I'd decouple the command creation a little bit.

Can we have:
```
// Hides the logic of appending commands
func (fake *FakeExec) AddCommand(cmd *FakeCmd) { ... }
```

And something else to create the command
```
// Hides the pain of creating commands
func NewMatchingArgsCommand(cmd string, args ...string) *FakeCmd {
    return func(cmd string, args ...string) exec.Cmd {
        // Some of the logic you have in your command
    }
}
```

Then it means that:
1. I can add command easily, but they don't have to follow your logic
2. I can invent new types of commands